File: backend/src/main/java/org/ucjgsalm/locagest/service/ConfigItemService.java
package org.ucjgsalm.locagest.service;

import jakarta.inject.Singleton;
import java.util.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.ucjgsalm.locagest.domain.*;
import org.ucjgsalm.locagest.domain.enums.*;
import org.ucjgsalm.locagest.repository.*;

/**
 * Gestion des items du catalogue et promotion des saisies libres.
 *
 * <p>Promotion LIBRE → catalogue : le resp. location ou le trésorier ajoute une saisie gardien
 * au catalogue. Cela crée un {@code ConfigItem} et met à jour la ligne existante (montant inchangé).
 */
@Slf4j
@Singleton
@RequiredArgsConstructor
public class ConfigItemService {

    private final LigneSejourRepository ligneRepo;
    private final ConfigItemRepository  itemRepo;

    /** Retourne uniquement les items actifs du catalogue. */
    public List<ConfigItem> listActifs() throws Exception {
        return itemRepo.findActifs();
    }

    /** Retourne tous les items du catalogue (actifs et inactifs). */
    public List<ConfigItem> listAll() throws Exception {
        return itemRepo.findAll();
    }

    /** Retourne toutes les saisies libres en attente de promotion (tous séjours). */
    public List<LigneSejour> listLignesLibres() throws Exception {
        return ligneRepo.findAllLibre();
    }

    /**
     * Promeut une saisie libre en item du catalogue.
     *
     * @param ligneId      ID de la {@code LigneSejour} LIBRE à promouvoir
     * @param categorie    catégorie catalogue (CASSE / LOCATION / SERVICE)
     * @param unite        unité de comptage
     * @param nomCatalogue désignation dans le catalogue (null = utilise la désignation de la ligne)
     * @return le {@code ConfigItem} créé
     * @throws NoSuchElementException  si la ligne n'existe pas
     * @throws IllegalStateException   si la ligne n'est pas LIBRE
     */
    public ConfigItem promouvoir(UUID ligneId,
                                 CategorieItem categorie,
                                 UniteItem unite,
                                 String nomCatalogue) throws Exception {
        var ligne = ligneRepo.findById(ligneId)
            .orElseThrow(() -> new NoSuchElementException("Ligne " + ligneId + " introuvable"));

        if (ligne.statut() != StatutLigne.LIBRE) {
            throw new IllegalStateException("La ligne " + ligneId + " n'est pas LIBRE (statut : " + ligne.statut() + ")");
        }

        var item = new ConfigItem(
            UUID.randomUUID(),
            nomCatalogue != null && !nomCatalogue.isBlank() ? nomCatalogue : ligne.designation(),
            categorie,
            ligne.prixUnitaire(),
            unite,
            true,
            null, null
        );
        var itemCree = itemRepo.insert(item);
        ligneRepo.promouvoir(ligneId, itemCree.id());

        log.info("Ligne {} promue en item catalogue '{}' (id={})", ligneId, itemCree.designation(), itemCree.id());
        return itemCree;
    }

    /**
     * Crée un nouvel item dans le catalogue.
     *
     * @param item item à créer (l'id sera remplacé par un UUID généré)
     */
    public ConfigItem creer(ConfigItem item) throws Exception {
        var withId = new ConfigItem(UUID.randomUUID(), item.designation(), item.categorie(),
            item.prixUnitaire(), item.unite(), item.actif(), null, null);
        log.info("Création item catalogue '{}' ({}/{})", withId.designation(), withId.categorie(), withId.unite());
        return itemRepo.insert(withId);
    }

    /**
     * Modifie un item existant du catalogue.
     *
     * @param item item avec les nouvelles valeurs (l'id doit correspondre à un item existant)
     */
    public ConfigItem modifier(ConfigItem item) throws Exception {
        log.info("Modification item catalogue {}", item.id());
        return itemRepo.update(item);
    }

    /**
     * Désactive logiquement un item.
     * Ne supprime jamais pour conserver l'historique des suppléments passés.
     */
    public void desactiver(UUID id) throws Exception {
        itemRepo.findById(id).ifPresent(item -> {
            var desactive = new ConfigItem(item.id(), item.designation(), item.categorie(),
                item.prixUnitaire(), item.unite(), false, item.createdAt(), item.updatedAt());
            try {
                itemRepo.update(desactive);
                log.info("Item catalogue {} désactivé", id);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        });
    }
}

File: plugin/includes/Service/ConfigItemService.php
<?php

declare(strict_types=1);

namespace Locagest\Service;

use Locagest\Repository\ConfigItemRepository;
use Locagest\Repository\LigneSejourRepository;
use Locagest\Utils\Exceptions\ConflictException;
use Locagest\Utils\Exceptions\InvalidInputException;
use Locagest\Utils\Exceptions\NotFoundException;

class ConfigItemService {

    public function __construct(
        private readonly ConfigItemRepository  $item_repo,
        private readonly LigneSejourRepository $ligne_repo,
    ) {}

    public function list_actifs(): array {
        return $this->item_repo->find_all_actifs();
    }

    public function create( array $data ): array {
        $this->validate( $data );
        $max_ordre = 0;
        foreach ( $this->item_repo->find_all_actifs() as $item ) {
            $max_ordre = max( $max_ordre, (int) $item['ordre'] );
        }
        $id = $this->item_repo->create( [
            'libelle'       => $data['libelle'],
            'categorie'     => $data['categorie'],
            'prix_unitaire' => $data['prix_unitaire'],
            'unite'         => $data['unite'] ?? 'UNITE',
            'ordre'         => $max_ordre + 1,
        ] );
        return $this->find_or_fail( $id );
    }

    public function update( int $id, array $data ): array {
        $this->find_or_fail( $id );
        $this->validate( $data, partial: true );
        $this->item_repo->update( $id, array_filter( [
            'libelle'       => $data['libelle']       ?? null,
            'categorie'     => $data['categorie']     ?? null,
            'prix_unitaire' => $data['prix_unitaire'] ?? null,
            'unite'         => $data['unite']         ?? null,
        ], fn( $v ) => $v !== null ) );
        return $this->find_or_fail( $id );
    }

    public function deactivate( int $id ): void {
        $this->find_or_fail( $id );
        $this->item_repo->deactivate( $id );
    }

    public function delete( int $id ): void {
        $this->find_or_fail( $id );
        if ( $this->item_repo->is_used( $id ) ) {
            throw new ConflictException( "Cet item est utilisé par des séjours existants. Vous pouvez uniquement le désactiver." );
        }
        $this->item_repo->hard_delete( $id );
    }

    /**
     * Promeut une saisie libre en item catalogue (effet atomique).
     * @param array $data { categorie_item: string, unite: string, nom_catalogue: string }
     */
    public function promouvoir( int $ligne_id, array $data ): array {
        $ligne = $this->ligne_repo->find_by_id( $ligne_id );
        if ( ! $ligne ) throw new NotFoundException( "Ligne #$ligne_id introuvable." );
        if ( $ligne['type_ligne'] !== 'LIBRE' ) throw new InvalidInputException( "Seules les lignes LIBRE peuvent être promues." );

        // Crée l'item catalogue
        $item_id = $this->item_repo->create( [
            'libelle'       => $data['nom_catalogue'] ?? $ligne['libelle'],
            'categorie'     => $data['categorie_item'],
            'prix_unitaire' => $ligne['prix_unitaire'],
            'unite'         => $data['unite'] ?? 'UNITE',
        ] );

        // Met à jour la ligne (montant inchangé)
        $this->ligne_repo->promouvoir( $ligne_id, $item_id );

        return $this->find_or_fail( $item_id );
    }

    private function find_or_fail( int $id ): array {
        $item = $this->item_repo->find_by_id( $id );
        if ( ! $item ) throw new NotFoundException( "Item catalogue #$id introuvable." );
        return $item;
    }

    private function validate( array $data, bool $partial = false ): void {
        $valid_categories = [ 'LOCATION', 'CASSE', 'INTERVENTION' ];
        $valid_unites     = [ 'UNITE', 'SEJOUR' ];

        if ( ! $partial || isset( $data['libelle'] ) ) {
            if ( empty( $data['libelle'] ) ) throw new InvalidInputException( 'Le libellé est obligatoire.' );
        }
        if ( ! $partial || isset( $data['categorie'] ) ) {
            if ( ! in_array( $data['categorie'] ?? '', $valid_categories, true ) ) {
                throw new InvalidInputException( 'Catégorie invalide. Valeurs acceptées : ' . implode( ', ', $valid_categories ) );
            }
        }
        if ( ! $partial || isset( $data['prix_unitaire'] ) ) {
            if ( ! isset( $data['prix_unitaire'] ) || $data['prix_unitaire'] < 0 ) {
                throw new InvalidInputException( 'Le prix unitaire doit être ≥ 0.' );
            }
        }
        if ( isset( $data['unite'] ) && ! in_array( $data['unite'], $valid_unites, true ) ) {
            throw new InvalidInputException( 'Unité invalide. Valeurs acceptées : ' . implode( ', ', $valid_unites ) );
        }
    }
}
